lib/weather/application.ex

```diff
@@ -7,6 +7,8 @@ defmodule Weather.Application do
 
   @impl true
   def start(_type, _args) do
+    :ok = Weather.Cache.start()
```

Comment:
Э-э-э-э… Почему вне дерева супервизоров? Кажется, это неправильно.

Reply:
Хм, наверное да. Переписал модуль, который содержит функции работы с кэшом через ETS с использованием генсервера и добавил его в дерево супервизоров.

Правда не очень пойму насколько это правильно, т.к. получается что не происходит работы со стэйтом самого генсервера в процессе работы с кэшом :thinking:

Reply:
Уточните, пожалуйста, что это значит? Вроде, по коду все ок, но там 37 файлов, я мог что-то пропустить.

Reply:
Конкретно речь о файле: `lib/weather/cache.ex`, т.е. сама имплементация кэша с использованием генсервера.

Возьмем такой пример (на скриншоте мой код в данный момент):
- стрелка 1 - мы просто "прокидываем" стэйт при ините процесса.
- стрелка 2 - так же просто прокидываем стэйт не используя его внутри самой функции, т.е. получается как будто используем генсервер и не пользуемся его состоянием.

Вопрос именно в этом. Не пойму допускается ли так делать. Ведь в доке к генсерверу написано, что:

> A GenServer is a process like any other Elixir process and **it can be used to keep state**, execute code asynchronously and so on.

![genserver_cache](https://user-images.githubusercontent.com/22988869/151698975-fed0502c-435b-4ab4-b70a-ace5b71397e5.jpg)

___

Т.е. как будто должно быть что-то такое:

```
  @impl true
  def handle_call({:get, key, ttl}, _from, state) do
    result =
      case :ets.lookup(@table, key) do
        [{^key, value, ts}] ->
          if current_timestamp() - ts <= ttl do
            value
          end

        _else ->
          nil
      end

    new_state = Map.put(state, key, result)

    {:reply, result, new_state}
  end
```

Reply:
> Абсолютно допустимо. Но можно и выбросить ets, и хранить кеш прямо в стейте
как мапу.

Это хорошо, спасибо за ответ! Оставлю с ETS.